### PR TITLE
Invalidate the CI cache when containers are changed

### PR DESCRIPTION
**Pull request by PureFunctor**
**Description of the change**

This changes the cache keys for the workflows to take into account the container they're being run on. As mentioned in https://github.com/purescript/purescript/pull/4391#issuecomment-1258910594, CI is currently failing with a `libtinfo` error:
```
/root/.stack/compiler-tools/x86_64-linux/ghc-9.2.3/bin/weeder: error while loading shared libraries: libtinfo.so.5: cannot open shared object file: No such file or directory
```

My guess is that it has to do with the cache being retained _after_ the container was changed since #4391 itself didn't run into this failure as it built its own `weeder`.

---

**Checklist:**

- [ ] Added a file to CHANGELOG.d for this PR (see CHANGELOG.d/README.md)
- [ ] Added myself to CONTRIBUTORS.md (if this is my first contribution)
- [ ] Linked any existing issues or proposals that this pull request should close
- [ ] Updated or added relevant documentation
- [ ] Added a test for the contribution (if applicable)

### CODE REVIEW

**Comment by PureFunctor** [comment]
It looks like my guess was correct 😄. Take a look at these runs:

pre-container-bump: https://github.com/purescript/purescript/actions/runs/2767436187/jobs/4349010103
container-bump-pr: https://github.com/purescript/purescript/actions/runs/3120713336/jobs/5061508296
post-container-bump: https://github.com/purescript/purescript/actions/runs/3132673764/jobs/5085265725

All of these runs use the same cache key for CI. If I understand correctly, the reason why the container-bump-pr was able to build just fine was because GitHub isolates the caches when on different branches or forks, so it just built its own `weeder`. When the PR was merged, post-container-bump now has access to the same cache that pre-container-bump did, causing the failure with `libtinfo`.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -84,7 +84,7 @@ jobs:
           path: |
             /root/.stack
             ${{ steps.haskell.outputs.stack-root }}
-          key: "${{ runner.os }}-MdyPsf-${{ hashFiles('stack.yaml') }}"
+          key: "${{ runner.os }}-${{ job.container.id }}-MdyPsf-${{ hashFiles('stack.yaml') }}"
```
[the remark this answers is not in the document]
Comment (PureFunctor) [reply .github/workflows/ci.yml:87]: Under "contexts" https://docs.github.com/en/actions/learn-github-actions/contexts

[the remark this answers is not in the document]
Comment (JordanMartinez) [reply .github/workflows/ci.yml:87]: Gotcha. Thanks!

**Review by JordanMartinez — APPROVED** [review]

**Comment by rhendric** [comment]
Nice, thanks for fixing this!

**Pull request merged**